pubsub/streaming-analytics/requirements.txt

```diff
@@ -0,0 +1 @@
+apache-beam==2.15.0
```

Comment:
Missing tests.

Reply:
@davidcavazos I couldn't quite figure out how to write tests. I tried a few different ways. Hope to get your help!

Reply:
Testing would also be easier if your `run` function had the signature `run(input_topic, output_path, window_size=1.0, pipeline_args=None)` and then `__main__` only parsed the arguments with `argparse`.

Optional, but I would also omit the `=` signs on the `pipeline_args` and break them into their own list element instead. I wander if there's a parsing error or something due to quoting or some other shell language assumption.

```
pipeline_args = [
  '--project', PROJECT,
  '--runner', 'DirectRunner',
  '--temp_location', 'gs://{}/pubsub/PubSubToGCS/temp'.format(BUCKET),
]
```

Also, can we use a local temporary directory or does it have to be GCS? (`'--temp_location', tempfile.mkdtemp()`).

For the `output_path`, you'll also want to make a unique directory so there are no collisions if this test is ran concurrently by more than one user. Something like `gs://{}/pubsub/PubSubToGCS/{}/output'.format(BUCKET, uuid.uuid4().hex)`.

Since this is a streaming job, it will run (and block) forever unless you stop it manually, so I would recommend [running on another Process with a timeout](https://dreamix.eu/blog/webothers/timeout-function-in-python-3). Do we need the `time.sleep(WAIT_TIME)`?

You probably also want to delete all the GCS resources created if the test succeeded.

As for your error, it's failing to decode the JSON messages in `AddTimestamps`. Could it be that you're passing arbitrary text rather than a JSON encoded message? Depending on your use case, maybe just converting the message into a string might be enough.